### ISSUE TITLE
Regression related to instance resolution

### ISSUE THREAD

**Issue by nad** [agda/agda#6700]
The following sequence of commands succeeds if Agda 2.6.4-9afe77020541b944331685ef2720a81bb312a925 is used, but fails with Agda-2.6.4-2af2ff114d167d9931034ca16d1bf6ba04efe661:
```
mkdir fresh && cd fresh
(git clone https://github.com/agda/agda-stdlib.git && cd agda-stdlib && git checkout 8c234b7ad402c3ca6ceb4e96e4b806a6575d71ac)
(git clone https://github.com/nad/equality && cd equality && git checkout 54bb9cacec5912b8bf77f42c2bf859dbda616252)
cat > libraries <<EOF
$PWD/agda-stdlib/standard-library.agda-lib
$PWD/equality/equality.agda-lib
EOF
cat > Bug.agda <<EOF
{-# OPTIONS --without-K --erasure #-}

open import Equality.Propositional
open import Tactic.By.Propositional

data Exp : Set where
  c : Exp → Exp → Exp

data Exp′ : Set where
  c′ : Exp′ → Exp′ → Exp′

record Rep (A : Set) : Set where
  field
    ⌜_⌝ : Exp → A

open Rep ⦃ … ⦄

⌜_⌝′ : Exp′ → Exp
⌜ c′ x y ⌝′ = c ⌜ x ⌝′ ⌜ y ⌝′

instance

  rep-Exp : ⦃ r : Rep Exp′ ⦄ → Rep Exp
  rep-Exp .⌜_⌝ x = ⌜ ⌜ x ⌝ ⌝′

  rep-Exp′ : Rep Exp′
  rep-Exp′ .⌜_⌝ = rep
    where
    rep : Exp → Exp′
    rep (c x y) = c′ (rep x) (rep y)

works : (x : Exp) → ⌜ ⌜ x ⌝ ⌝′ ≡ ⌜ x ⌝
works x = works′ x
  where
  works′ : (x : Exp) → ⌜ ⌜ x ⌝ ⌝′ ≡ ⌜ x ⌝
  works′ (c x y) =
    c ⟨ ⌜ ⌜ x ⌝ ⌝′ ⟩ ⌜ ⌜ y ⌝ ⌝′  ≡⟨ ⟨by⟩ (works x) ⟩
    c ⌜ x ⌝ ⟨ ⌜ ⌜ y ⌝ ⌝′ ⟩       ≡⟨ ⟨by⟩ (works y) ⟩∎
    c ⌜ x ⌝ ⌜ y ⌝                ∎

fails : (x : Exp) → ⌜ ⌜ x ⌝ ⌝′ ≡ ⌜ x ⌝
fails (c x y) =
  c ⟨ ⌜ ⌜ x ⌝ ⌝′ ⟩ ⌜ ⌜ y ⌝ ⌝′  ≡⟨ ⟨by⟩ (fails x) ⟩
  c ⌜ x ⌝ ⟨ ⌜ ⌜ y ⌝ ⌝′ ⟩       ≡⟨ ⟨by⟩ (fails y) ⟩∎
  c ⌜ x ⌝ ⌜ y ⌝                ∎
EOF
agda --no-default-libraries --library-file=libraries -lstandard-library -lequality -i. Bug.agda
```
The problem seems to have been introduced by "[ fix #6364 ] Only filter instances based on their type, not their body" (2af2ff114d167d9931034ca16d1bf6ba04efe661), but perhaps the underlying problem is older.

I'm not sure that this is a bug, but I don't see why there should be a difference between `works` and `fails`:

* Agda fails to infer some things in `fails` that are inferred in `works`.
* The only difference is that `works` uses a local helper function, with the same type as `works`, and the recursive calls use the top-level function instead of the local one.
 
@jespercockx, can you investigate?

**Comment by jespercockx** [comment]
> (git clone https://github.com/nad/equality && cd equality && git checkout 54bb9cacec5912b8bf77f42c2bf859dbda616252)

@nad this commit does not seem to exist.

**Comment by nad** [comment]
Sorry, I have pushed the missing commit now. @jespercockx, I think it would be good if you could take a look at this before the release.

**Comment by jespercockx** [comment]
I'm at IFL this week so I might only have time to look at this when I'm back on Friday.

**Comment by jespercockx** [comment]
The problem here seems to be caused by the occurs checker, specifically the check for avoiding cyclic solutions. In both `succeeds` and `fails`, Agda gets to the point that the macro finds a solution and tries to use it to instantiate the hole. But in the case of `fails` this fails because the solution contains `fails`, which contains the metavariable itself, which is not allowed. But adding the extra layer of indirection is seemingly enough to fool the occurs checker into thinking the solution is non-circular.

I am actually not really sure why we have this check in the occurs checker the first place. To me it seems perfectly fine to solve a metavariable with a solution that contains the function in whose body the meta appears, since this is just a recursive call to the function. If the definition is indeed cyclic, then this should be caught by the termination checker, not by the occurs checker. So we could experiment with simply removing this check, but perhaps right before the release is not the right moment.

[I am not sure why my specific change to the instance search has triggered this behaviour, but the example involves a lot of imports AND instance arguments AND tactics which makes it very hard to understand what's going on or to minimize it.]

**Comment by jespercockx** [comment]
Here is a small example where the current check seems to restrictive to me:
```agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality

mutual
  foo : Nat → Nat
  foo zero = zero
  foo (suc n) = suc _

  test : (n : Nat) → foo (suc n) ≡ suc (foo n)
  test n = refl
```
Agda complains:
```
Failed to solve the following constraints:
  _n_6 (n = n) = foo n : Nat (blocked on _n_6)
```
but of course just manually replacing the `_` with `foo n` is perfectly fine. 

@andreasabel do you have an idea of why we have this check? If not, I would propose that we leave this issue as it is for the release of 2.6.4 (as it is actually just an irrelevant change that triggers a pre-existing limitation/bug) and experiment with removing the check for the next release.

**Comment by jespercockx** [comment]
Here is the variant of my example that circumvents the check and *is* accepted:
```agda

open import Agda.Builtin.Nat
open import Agda.Builtin.Equality

bar : Nat → Nat
bar n = bar' n
  where
  bar' : Nat → Nat
  bar' zero = zero
  bar' (suc n) = suc _

  tester : (n : Nat) → bar' (suc n) ≡ suc (bar n)
  tester n = refl
```
The reason why it is accepted is because at the time of solving the meta, the clause `foo n = foo' n` of `foo` is still being checked and hence hasn't been added to the signature, so the occurs checker does not see any circularity.

Some git archaeology leads to issue #585 where this behavior was originally introduced 11 years ago by @andreasabel in response to a bug reported by @nad . The check is trying to avoid non-termination during the normalization step of the occurs checker. In this light it makes sense that the directly recursive definition is rejected but the one with the indirection is accepted: in the latter case, the clause has not yet been added so there is no risk of an infinitely unfolding meta solution. 

Here is another fun example to play with:
```agda
mutual
  foo : Nat → Nat
  foo' : Nat → Nat
  test : (n : Nat) → foo n ≡ foo' n

  foo n = _
  test n = refl
  foo' n = foo n
```
Depending on the order of the clauses of `test` and `foo'`, either the meta gets solved and you get a termination error, or the meta is left unsolved.

In conclusion, there is actually a good justification for Agda's behavior here but the reasoning is so arcane and depends on implementation details of Agda (in particular, the time when clauses are added to the signature and the way Agda decides whether to normalize terms during occurs checking) that it is very hard if not impossible to explain to the user what goes wrong. I'm not sure what would be a good solution here, but I'll think about it. Either way, I don't think this is something that we can reasonably expect to fix before 2.6.4.

**Comment by nad** [comment]
> In this light it makes sense that the directly recursive definition is rejected but the one with the indirection is accepted: in the latter case, the clause has not yet been added so there is no risk of an infinitely unfolding meta solution.

I seem to recall that there is some machinery for marking code as unsafe to unfold before the termination checker has accepted it. However, I think that this machinery is only applicable to definitions, not expressions, so presumably it would not help in this case.

**Comment by nad** [comment]
See also #7847.
